src/desktop/window.rs

```diff
@@ -0,0 +1,312 @@
+    /// surface.
+    pub fn surface_under<P: Into<Point<f64, Logical>>>(
+        &self,
+        point: P,
+    ) -> Option<(wl_surface::WlSurface, Point<i32, Logical>)> {
+        let point = point.into();
+        if let Some(surface) = self.0.toplevel.get_surface() {
+            for (popup, location) in PopupManager::popups_for_surface(surface).ok().into_iter().flatten() {
+                if let Some(result) = popup.get_surface().and_then(|surface| under_from_surface_tree(surface, point, location)) {
+                    return Some(result);
+                }
+            }
+
+            under_from_surface_tree(surface, point, (0, 0))
+        } else { None }
+    }
+
+    /// Damage of all the surfaces of this window
+    pub fn accumulated_damage(&self, for_values: Option<(&Space, &Output)>) -> Vec<Rectangle<i32, Logical>> {
+        let mut damage = Vec::new();
+        if let Some(surface) = self.0.toplevel.get_surface() {
+            damage.extend(damage_from_surface_tree(surface, (0, 0), for_values).into_iter().flat_map(|rect| rect.intersection(self.bbox())));
+            for (popup, location) in PopupManager::popups_for_surface(surface)
+                .ok()
+                .into_iter()
+                .flatten()
+            {
+                if let Some(surface) = popup.get_surface() {
+                    let bbox = bbox_from_surface_tree(surface, location);
+                    let popup_damage = damage_from_surface_tree(surface, location, for_values);
+                    damage.extend(popup_damage.into_iter().flat_map(|rect| rect.intersection(bbox)));
+                }
+            }
+        }
+        damage
+    }
+
+    pub fn toplevel(&self) -> &Kind {
+        &self.0.toplevel
+    }
+
+    pub fn user_data(&self) -> &UserDataMap {
+        &self.0.user_data
+    }
+}
+
+pub fn draw_window<R, E, F, T, P>(
+    renderer: &mut R,
+    frame: &mut F,
+    window: &Window,
+    scale: f64,
+    location: P,
+    damage: &[Rectangle<i32, Logical>],
+    log: &slog::Logger,
+) -> Result<(), R::Error>
+where
+    R: Renderer<Error = E, TextureId = T, Frame = F> + ImportAll,
+    F: Frame<Error = E, TextureId = T>,
+    E: std::error::Error,
+    T: Texture + 'static,
+    P: Into<Point<i32, Logical>>,
+{
+    let location = location.into();
+    if let Some(surface) = window.toplevel().get_surface() {
+        draw_surface_tree(renderer, frame, surface, scale, location, damage, log)?;
+        for (popup, p_location) in PopupManager::popups_for_surface(surface)
+            .ok()
+            .into_iter()
+            .flatten()
+        {
+            if let Some(surface) = popup.get_surface() {
+                let damage = damage.iter().cloned().map(|mut geo| { geo.loc -= p_location; geo }).collect::<Vec<_>>();
+                draw_surface_tree(renderer, frame, surface, scale, location + p_location, &damage, log)?;
```

Comment:
I'm not sure if location of popups here is correct.
![example)](https://user-images.githubusercontent.com/20758186/147796160-f01faf22-1e0c-46f7-bdc3-f9a219f2c096.png)

In Anodium I do: `location + window.geometry().loc + p_location`, but I'm not sure if that's correct either, as you can see in the screenshot it is also a little bit off in comparison to gnome.